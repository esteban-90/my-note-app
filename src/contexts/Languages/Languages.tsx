import type { FC, PropsWithChildren } from 'react'
import type { Language, LanguageUtils } from '@/types'
import { createContext, useContext } from 'react'
import useLocalStorage from 'use-local-storage'
import { IntlProvider } from 'react-intl'
import { makeKey } from '@/helpers'
import { translationEN_US, translationES_PY } from '@/translations'

const defaultValues: LanguageUtils = {
  language: navigator.language as Language,
  changeLanguage: () => undefined,
}

const context = createContext<LanguageUtils>(defaultValues)

export const LanguageProvider: FC<PropsWithChildren> = ({ children }) => {
  const defaultLanguage = defaultValues.language
  const [language, setLanguage] = useLocalStorage<Language>(makeKey('language'), defaultLanguage)

  const changeLanguage = () => {
    setLanguage((previousLanguage) => (previousLanguage === 'en-US' ? 'es-PY' : 'en-US'))
  }

  const { Provider } = context
  const utils: LanguageUtils = { language, changeLanguage }
  const messages = { 'en-US': translationEN_US, 'es-PY': translationES_PY }

  return (
    <Provider value={utils}>
      <IntlProvider locale={language} defaultLocale={defaultLanguage} messages={messages[language]}>
        {children}
      </IntlProvider>
    </Provider>
  )
}

export const useLanguage = () => useContext<LanguageUtils>(context)
