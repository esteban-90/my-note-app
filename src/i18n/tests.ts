import type { Resource } from 'i18next'
import { createInstance } from 'i18next'
import { initReactI18next } from 'react-i18next'
import { ns, supportedLngs } from './app'

const resources: Resource = {}
export const i18nInstance = createInstance()

void (async () => {
  for (const n of ns) {
    for (const lng of supportedLngs) {
      resources[lng] ??= {}
      resources[lng] = {
        ...resources[lng],
        [n]: await import(`../../public/translations/${lng}/${n}.json`),
      }
    }
  }
})()

i18nInstance.use(initReactI18next).init({
  ns,
  supportedLngs,
  fallbackLng: supportedLngs[0],
  lng: supportedLngs[0],
  react: { useSuspense: false },
  resources,
})
