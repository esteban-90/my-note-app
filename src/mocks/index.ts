import type { Note, Language, Theme } from '@/types'
import uniqid from 'uniqid'
import { makeKey } from '@/helpers'

let seconds = 0

export const makeMockNote = (content: string): Note => {
  return {
    id: uniqid('note-'),
    content,
    createdAt: new Date(1995, 11, 17, 3, 24, seconds++).toISOString(),
  }
}

const makeMockNotes = (quantity = 1): Note[] => {
  return new Array(quantity).fill(null).map((_, index) => makeMockNote(`Hello test ${index + 1}\nTesting ${index + 1}`))
}

export const mockNotes = makeMockNotes(5)

export const mockNoteKey = makeKey('notes')
export const mockLanguageKey = makeKey('language')
export const mockThemeKey = makeKey('theme')

export const mockENLanguage: Language = 'en-US'
export const mockESLanguage: Language = 'es-PY'

export const mockDayTheme: Theme = 'day'
export const mockNightTheme: Theme = 'night'

export const formatMockDate = (date: string, locale: Language = 'en-US') => {
  return new Date(date).toLocaleString(locale, { dateStyle: 'long', timeStyle: 'medium' })
}
