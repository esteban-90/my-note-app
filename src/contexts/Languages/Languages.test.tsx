import { mockLanguageKey, mockENLanguage, mockESLanguage } from '@/mocks'
import { screen, render, fireEvent, configure } from '@/tests'
import { LanguageProvider, useLanguage } from './Languages'

describe(LanguageProvider.name, () => {
  let asFragment: () => DocumentFragment

  const Child = () => {
    const { language, changeLanguage } = useLanguage()
    return (
      <div role='presentation'>
        <p>{language}</p>
        <button onClick={changeLanguage}>Change language</button>
      </div>
    )
  }

  const UI = (
    <LanguageProvider>
      <Child />
    </LanguageProvider>
  )

  beforeAll(() => {
    configure({ throwSuggestions: true })
  })

  beforeEach(() => {
    ;({ asFragment } = render(UI))
  })

  it('should match snapshot', () => {
    expect(asFragment()).toMatchSnapshot()
  })

  it('should provide a specific language at a time to child elements and save it to storage', () => {
    const childElement = screen.getByText(mockENLanguage)
    expect(childElement).toBeInTheDocument()

    const savedLanguage = localStorage.getItem(mockLanguageKey)
    expect(savedLanguage).not.toBeNull()
    expect(savedLanguage).not.toContain(mockESLanguage)
    expect(savedLanguage).toContain(mockENLanguage)
  })

  it('should be able to change language and save it to storage with the new value', () => {
    const buttonElement = screen.getByRole('button', { name: /^change language/i })
    fireEvent.click(buttonElement)

    const childElement = screen.getByText(mockESLanguage)
    expect(childElement).toBeInTheDocument()

    const savedLanguage = localStorage.getItem(mockLanguageKey)
    expect(savedLanguage).not.toBeNull()
    expect(savedLanguage).not.toContain(mockENLanguage)
    expect(savedLanguage).toContain(mockESLanguage)

    localStorage.removeItem(mockLanguageKey)
  })
})
