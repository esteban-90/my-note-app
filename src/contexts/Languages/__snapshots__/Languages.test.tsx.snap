// Vitest Snapshot v1

exports[`LanguageProvider > should match snapshot 1`] = `
<DocumentFragment>
  <div
    role="presentation"
  >
    <p>
      en-US
    </p>
    <button>
      Change language
    </button>
  </div>
</DocumentFragment>
`;
